### DISCUSSION TITLE
extend schema during runtime

### DISCUSSION THREAD

**Discussion by roderm** [99designs/gqlgen#1885] [Ideas]
### What happened?
With GraphQL SDL it's possible to extend a schema, this works well if the extensions are given at compile time. If e.g. as plugin is loaded during runtime, the Schema can be extended, but resolvers can't be added:
```golang
schema := generated.NewExecutableSchema(generated.Config{Resolvers: &graph.Resolver{}})
schemaType := schema.Schema().Types["Todo"]
schemaType.Fields = append(schemaType.Fields, &ast.FieldDefinition{
	Name: "github_issue",
	Type: ast.NamedType("String", nil),
})
schemaType := schema.Schema().Types["NewTodo"]
schemaType.Fields = append(schemaType.Fields, &ast.FieldDefinition{
	Name: "github_issue",
	Type: ast.NamedType("String", nil),
})
```
* introspection shows the fields
* for Inputs the fields are just ignored
* querying the field throws "unknown field github_issue" from resolving

#### Mutation
Getting the added field over the passed in Context. This would also only work for Types defined in the original schema oder builtin types. Else a `map[string]interface{}` must be used.
```go
func (r *mutationResolver) CreateTodo(ctx context.Context, input model.NewTodo) (*model.Todo, error) {
	rawInput := graphql.CollectInputs(ctx)
	...
}
```
#### Query
Adding custom resolver to the Schema (`executableSchema`/`executionContext`) to fill manually:
```go
schema := generated.NewExecutableSchema(generated.Config{Resolvers: &resolver})
schemaType, ok := schema.Schema().Types["Todo"]
schemaType.Fields = append(schemaType.Fields, 
	&ast.FieldDefinition{
		Name: "github_url",
		Type: ast.NamedType("String", nil),
	},
)
schema.AddResolver("Todo.github_url", func(ctx context.Context, field graphql.CollectedField, obj *model.Todo) graphql.Marshaler {
	return graphql.MarshalString("https://github.com/....")
})
```
#### Merge Schema
To keep comfort of the generation but still being able to extend during runtime, a merge during runtime would be nice:
```go
schema1 := generated.NewExecutableSchema(generated.Config{Resolvers: &resolver})
schema2 := generatedExt.NewExecutableSchema(generatedExt.Config{Resolvers: &resolverExt})
schema.Extend(schema2)
```

### versions
 - `gqlgen version`? v0.16.0
 - `go version`? 1.17

Would love to contribute this feature, but currently no idea where to begin. Is this use case reasonable and wanted by others as well?

**Comment by roderm** [comment]
I've created a [protoype](https://github.com/roderm/gqlgen/tree/rot/extend-schema) for this feature. With this it's possible to use `schema1.Extend(schema2)`, but complexity and tests are not yet implemented. 
Would love some feedback.

**Reply by magi-wang** [reply]
Great job ~

**Comment by khanakia** [comment]
@roderm was this feature ever merged to main repo ?

**Comment by roderm** [comment]
@khanakia it has not been merged. Also I tried to solve this use-case partly with federation, I still had to implement some logic for inputs in this library.
You can find them here: https://github.com/roderm/gqlgen/commits/rot/schema/
* The basic idea on this is to add an `extend input` SDL to the schema and register and `afterUnmarshal` function to catch the other fields. This is used for filter, order and batch inputs as their is no apollo federation solution for this yet.
* Additionaly a graphQL service created with [a fork of graphql-go](https://github.com/roderm/graphql-go/tree/federation-subgraph), I extend the inputs with federation and pack both services behind a [graphql-go-tools gateway](https://github.com/wundergraph/graphql-go-tools/)

**Reply by khanakia** [reply]
Got it

**Reply by roderm** [reply]
If I find the time to create the tests for the "input-only" extension, I'll try to get a PR here.
